src/main/java/htsjdk/tribble/index/IndexFactory.java

```diff
@@ -73,47 +78,43 @@
      */
     public enum IndexType {
-        LINEAR(LinearIndex.MAGIC_NUMBER, LinearIndex.INDEX_TYPE, LinearIndexCreator.class, LinearIndex.class, LinearIndexCreator.DEFAULT_BIN_WIDTH),
-        INTERVAL_TREE(IntervalTreeIndex.MAGIC_NUMBER, IntervalTreeIndex.INDEX_TYPE, IntervalIndexCreator.class, IntervalTreeIndex.class, IntervalIndexCreator.DEFAULT_FEATURE_COUNT),
+        LINEAR(AbstractIndex.MAGIC_NUMBER, LinearIndex.INDEX_TYPE, true, LinearIndex::new, LinearIndexCreator.DEFAULT_BIN_WIDTH),
+        INTERVAL_TREE(AbstractIndex.MAGIC_NUMBER, IntervalTreeIndex.INDEX_TYPE, true, IntervalTreeIndex::new, IntervalIndexCreator.DEFAULT_FEATURE_COUNT),
         // Tabix index initialization requires additional information, so generic construction won't work, thus indexCreatorClass is null.
-        TABIX(TabixIndex.MAGIC_NUMBER, null, null, TabixIndex.class, -1);
+        TABIX(TabixIndex.MAGIC_NUMBER, null, false, TabixIndex::new, -1) ;
 
         private final int magicNumber;
         private final Integer tribbleIndexType;
-        private final Class<IndexCreator> indexCreatorClass;
         private final int defaultBinSize;
-        private final Class<Index> indexType;
+        private final boolean canCreate;
+        private final IndexFromStreamFunction createFromInputStream;
 
         public int getDefaultBinSize() {
             return defaultBinSize;
         }
 
-        public IndexCreator getIndexCreator() {
-            try {
-                return indexCreatorClass.newInstance();
-            } catch ( final InstantiationException | IllegalAccessException e ) {
-                throw new TribbleException("Couldn't make index creator in " + this, e);
-            }
+        public boolean canCreate() {
+            return canCreate;
         }
 
-        public boolean canCreate() {
-            return indexCreatorClass != null;
+        private interface IndexFromStreamFunction {
+            Index apply(InputStream t) throws IOException;
         }
 
-        IndexType(final int magicNumber, final Integer tribbleIndexType, final Class creator, final Class indexClass, final int defaultBinSize) {
+        IndexType(final int magicNumber, final Integer tribbleIndexType, boolean canCreate, IndexFromStreamFunction createFromInputStream, final int defaultBinSize) {
             this.magicNumber = magicNumber;
             this.tribbleIndexType = tribbleIndexType;
-            indexCreatorClass = creator;
-            indexType = indexClass;
+            this.canCreate = canCreate;
+            this.createFromInputStream = createFromInputStream;
             this.defaultBinSize = defaultBinSize;
         }
 
         public Integer getTribbleIndexType() {
             return tribbleIndexType;
         }
 
-        public Class<Index> getIndexType() {
-            return indexType;
+        public Index createIndex(InputStream in) throws IOException{
```

Comment:
It seems like this should reject creation of a TABIX index (honor `canCreate()`) since it doesn't have the sequence dictionary? I think the previous (index creator) method would have blown up on TABIX, but now I'm not sure what will happen with this one. Also, since its public perhaps it should wrap IOException with RuntimeIOException, and `final` on arg.

Reply:
I don't really understand what `canCreate()` is supposed to mean.  You can totally create a tabixIndex from a stream without providing a sequence dictionary.  The previous method that used getIndexType() worked on tabix as far as I can see and didn't check canCreate().   I suspect that canCreate may just be completely outdated and irrelevant, but I didn't want to figure it out as part of stripping out the bad reflection.

Reply:
Ok, fair enough.